### ISSUE TITLE
Pull Request merging is opaque and unpredictable

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#5986]
Getting pull requests merged is not as predicatable as it should be, in particular, there is no reliable way for a PR author to know if their PR will be merged and when. This is a large drag of time, and difficults planning work. In particular, we'd need:

- To expose the filter used by the RMs to consider candidate PRs. This way, authors could check on their own if some step is missing.
- To provide an ordering mechanism when several PRs may conflict.
- To provide predicatable merge windows ["week 52, merges will happen in Tue/Wed at 10"] so authors don't have to be constantly checking for notifications.

**Comment by maximedenes** [comment]
Indeed, now is a good time to make progress on this. I can start by sharing the filter I currently use:

https://github.com/coq/coq/pulls?utf8=%E2%9C%93&q=is%3Apr%20is%3Aopen%20-label%3A%22needs%3A%20discussion%22%20-label%3A%22needs%3A%20testing%22%20-label%3A%22needs%3A%20fixing%22%20-label%3A%22needs%3A%20progress%22%20-label%3A%22needs%3A%20rebase%22%20-label%3A%22needs%3A%20review%22%20-label%3A%22needs%3A%20help%22%20-label%3A%22needs%3A%20independent%20fix%22%20-review%3Achanges_requested

I will add it to the contributing guide and keep it up to date.

Can you explain what kind of things you mean by "an ordering mechanism"? I had the feeling that planning conflicting PRs was complex enough that it would be hard to write systematic rules, but you may have something in mind.

As for the merge windows, I'm ok with it, I'll think a bit about the planning.

**Comment by ejgallego** [comment]
Thanks @maximedenes ; I have nothing too concrete in mind for "ordering", I guess we should collectively brainstorm a bit. In some cases maybe when the RM do detect conflicts they could just indicate in the corresponding PRs "this will go after that".

I dunno, let's hear what others do think.

**Comment by ejgallego** [comment]
> I can start by sharing the filter I currently use:

Interesting, the filter picks some unexpected PRs such as #1066. I see also you don't rely on the CI, but that gives good visual feedback anyways [on Gitlab you get much better immediate display for failure AFAICT]

**Comment by ejgallego** [comment]
As for merge windows, I'll personally be happy with a "once a week" guaranteed window, with of course, more merging at the discretion of RM.

I like that determinacy because I could then schedule rebases after the windows, whereas now it is a pain to have to rebase almost randomly.

**Comment by maximedenes** [comment]
> I see also you don't rely on the CI

That's a feature, not a bug. When a PR fails CI testing, I usually inspect the results and only if I find it is a non spurious failure (random problems in our testing, known upstream breakage), I put a "needs: fixing" label which this filter uses.

**Comment by Zimmi48** [comment]
Thanks for sharing your filter @maximedenes. I see that you remove PRs that have "Change requested". This is fine but it is good to know: this means that if someone requests a change but then is unresponsive after our update, we shouldn't hesitate to dismiss their review. In fact, this requirement could probably be upgraded to "Approved", but that would require everyone to start using the "Approve" feature (also note that for a PR to be in "Approved" state it needs to be approved by one of the people having write rights, but a lot of people have these rights nowadays).

Also note that you can add `base:master` in your filter (since you don't merge PRs on stable branches). This would not change the list most of the time, but could still be useful.

I would also benefit from knowing when are your merge windows and would even encourage you to have these in the mornings (if you do not have any particular preference), because I usually backport PRs in the afternoon. Sometimes, I start backporting and then notice that you have merged more PRs. This is not ideal.

**Comment by herbelin** [comment]
@maximedenes: Do you consider to be part of your role to actively watch the well progressing of all PRs, compared to watch only those corresponding to a fixed filter?

Doing this would be I think _extremely valuable_. Watching actively is taking a lot of time, even if only for one's own PRs and I suspect that this is part of the concerns of @ejgallego.

I'm trying to be active on all PRs from time to time, @Zimmi48 is also doing it regularly, and I can see how difficult it is. So, you, would you be ready to engage, as part of your role, to ensure that no PR remains in the pipeline for more than "necessary"?

To avoid the overheads of rebasing and rebasing, of watching and re-watching, of switching context again and again, I think that the number of PRs being processed should remain much lower than it is now. I would roughly tend to have no more than 30 live PRs at the same time, excluding deliberatedly experimental PRs such as #140, but including the heavy PRs such as #268, #930, #982, #991, and we crucially need someone to play this role.

What do you think?

**Comment by ejgallego** [comment]
I wonder how to make more progress on this. Maybe the RMs should outline a plan of action?

**Comment by ejgallego** [comment]
> communicating expected merge windows should be easy steps.

That would really help, now we have to check every once in a while whether our PRs have been merged and this is kinda of a pain. Would be much better if we could plan that.

Even you could set the gitter's channel topic with that information "[next MW: 20/11/2017 at 15]"

**Comment by maximedenes** [comment]
> Do you consider to be part of your role to actively watch the well progressing of all PRs, compared to watch only those corresponding to a fixed filter?

That's already what I usually do, but of course I often miss some (like others do).

> So, you, would you be ready to engage, as part of your role, to ensure that no PR remains in the pipeline for more than "necessary"?

I'm not sure I fully understand what that would mean, but from what I understand, I would say no, I can't guarantee that, sorry.

**Comment by maximedenes** [comment]
> Even you could set the gitter's channel topic with that information "[next MW: 20/11/2017 at 15]"

Yeah, that's an interesting idea. It would be less invasive than sending e-mails.

**Comment by herbelin** [comment]
@maximedenes: Thanks a lot for your answer. What you're doing for the team and the community is very useful and it would be a pity that further human-being-to-human-being discussions where you  explain clearly the role that you want to play could not eventually happen.

**Comment by Zimmi48** [comment]
So this morning was one of your merge windows and you merged a few PRs but your list still contains 21 opened PRs (after adding the additional ` -label:"needs: feedback"  -label:"help wanted" -status:pending base:master` filters, which remove respectively 4, 3, 9 and 1 PRs).

How do you manage such a long list? I know many of these PRs are still actively discussed and thus not ready to merge yet, but it is not obvious which ones when just looking at the list.

If I were you, I would also add `sort:updated-asc` to the search. Then you suddenly see that among these 21 PRs, there is [one](https://github.com/coq/coq/pull/370) whose last update was 15 days ago and [another one](https://github.com/coq/coq/pull/6150) whose last update was 8 days ago. Both are actually waiting on some other PRs getting merged first but have no flag to indicate this.

Then there is [one](https://github.com/coq/coq/pull/6139) where @ppedrot and you commented to ask for changes, but because none of you used the "Request changes" GitHub feature, the PR is still showing in this list.

Then there is [a PR](https://github.com/coq/coq/pull/6192) which was last updated two days ago by you saying "Will merge once Travis is happy." and it is rare enough to note that this PR has a green mark (Travis *and* Appveyor are happy). So why didn't you merge this one this morning? I suppose that you just missed it because skimming through this long list is just too difficult. The [following PR](https://github.com/coq/coq/pull/6187) is also in a mergeable state and has passed all checks. The [next PR](https://github.com/coq/coq/pull/6189) is also in a mergeable state (Travis has failed but only because GeoCoq timed out, damn!).

Then there are two PRs with questions not yet answered even if the last comment was two days ago (incidentally I've just added `needs: rebase` to both of these), then there is [a trivial clean-up](https://github.com/coq/coq/pull/6200) that received no comment at all yet and has passed all checks.

I will stop here but I hope you get my point (and my question), do you?

**Comment by maximedenes** [comment]
In fact, I realize now I think I cannot announce merge windows reliably.

**Comment by maximedenes** [comment]
> Both are actually waiting on some other PRs getting merged first but have no flag to indicate this.

@Zimmi48 what is the proper label to indicate this?

**Comment by Zimmi48** [comment]
> what is the proper label to indicate this?

We should probably create a new one, no? I just don't know how to call it.

**Comment by JasonGross** [comment]
Suggestions: `needs: external merge`, `needs: merge of dependent PR`

**Comment by maximedenes** [comment]
Thanks for the suggestions, I'll try to implement it next time I need it.

@ejgallego I put my filter in the contributing guide, and already updated it a couple of times. I tried announcing merge windows but stopped because I can't do it reliably.

Is there any other concrete action that should be taken or can I close this issue for now?

**Comment by ejgallego** [comment]
I feel like we need to refine the process more, but I am out of ideas for the moment. A last question, do we have a "before merge PR" checklist?

If so, we can close this IMHO.

**Comment by maximedenes** [comment]
> A last question, do we have a "before merge PR" checklist?

We don't, and I'd say addressing this is part of the issue on sharing our RM scripts/processes. (there is another open issue for that)

**Comment by maximedenes** [comment]
I fully agree, and I do plan to publish that information. But if that's ok, I'll close this issue since the remainder seems redundant with #5976, at least for now.

**Comment by herbelin** [comment]
May I formulate a suggestion? Would it be possible @maximedenes, that you announce say 24 or even 12 hours in advance what you are going to merge?

For instance, you use your filter, you see that this one has a travis problem and you write a comment about it. Then you look at another PR, you want to review it and you do it. And then you see that a third PR seems ready to you and you want to merge it. Instead of merging it at once, you write something of the following form: "This PR seems mergeable to me. I will do it tomorrow at 10am if no opposition by then.".

What do you think of following such kind of protocol?

**Comment by maximedenes** [comment]
Like I said above, I can't do that because if I need to work on a task which has higher priority (typically for the Consortium), I won't be able to merge "tomorrow at 10 am".

More generally, I don't see the point in delaying PR merges, and @ejgallego explicitly said he was opposed to it during the last WG.

**Comment by Zimmi48** [comment]
I also already expressed that I thought that if the protocol for merging must include a time-span before a PR can be merged, this time-span should start when the PR was opened, not when it is ready. Given than today, Travis is taking about 3 days to build a PR, this minimal time period looks like it is respected.

@herbelin What do you think of this solution instead? When you see a PR that you think you want to have a look at before it is merged, you go to this PR and say "I want to review this before it is merged, but I won't be able to do it before *this date*.", then you self-request a review not to forget, and possibly you add the `needs: review` label to clearly express to Maxime not to merge this.

**Comment by maximedenes** [comment]
And let's remind everybody that in the unlikely event that I merge very quickly a PR that is not controversial, it is fine to still discuss its integration. If it turns out we finally don't want to integrate it, we can revert it. The only thing we need to make sure is that it doesn't happen often. All this was already said during the last WG.

**Comment by ejgallego** [comment]
Yeah, reverting is cheap, delaying is not.

**Comment by ejgallego** [comment]
We should not fear revert.

**Comment by herbelin** [comment]
@maximedenes: saying "tomorrow at 10 am" was just an example. If you feel more comfortable with "at sometimes tomorrow", "at some time in the week", "the next time I'll be working at PRs", etc. this would be perfectly fine.

The idea was just to make it more predictable, less arbitrary, and that noone, like me, feels that the discussion has been bypassed.

But of course, this has to be compatible with your own organisation of time.

@ZiimmI48: The problem I see is that the comments on PRs are not independent. There are true discussions and interactions between the comments. It is frustrating to see a comment and the discussion closed without notice while you wanted to add something to this comment. And sometimes, you think that someone else should say something but the someone else does not say something, so you're postponing the time of saying yourself something and the absence of a limit before which you know that you have to say what you want to say is a problem in this case.

And after all, isn't it the discussion which is in fact more important that the merge itself? In particular reverting would not replace a missing discussion. If it were only about merging, we could merge by default every travis-ok PR and discuss and amend afterwards, as we did before.

> Yeah, reverting is cheap, delaying is not.

As already said, I think that 12 hours is peanut compared to the average lifespan of a PR. But the benefits on the feeling that time for discussion has been respected are immense.

**Comment by maximedenes** [comment]
I'm not sure I follow. In the scenarios you mention, why doesn't the PR have the "needs: discussion" label?

**Comment by Zimmi48** [comment]
I think @herbelin here is talking about discussion in a broader sense.

**Comment by Zimmi48** [comment]
@maximedenes has already been posting comments such as "Will merge once Travis is happy." so you are just asking a generalization of this in the form of "Will merge in my next batch of merges." Then I say, why not. BTW @maximedenes you could use GitHub "save replies" feature to avoid the hassle of having to type this each time.

**Comment by Zimmi48** [comment]
You could even use a GitHub search then to find which PRs you said you would merge. If the PR isn't showing with a blue mark (meaning there were more comments), then you proceed.

**Comment by maximedenes** [comment]
I stopped doing batches of merges (though it can still happen after the week-end or the night). I now merge them on-the-fly. IMHO it is up to authors and developers to signal that they want more discussion (using labels).

**Comment by Zimmi48** [comment]
> I stopped doing batches of merges (though it can still happen after the week-end or the night). I now merge them on-the-fly.

I remarked this. IMHO your way of doing batches of merges was one of the main points in favor of having a single person do the merges (and it reduces Travis load).

**Comment by maximedenes** [comment]
I can switch back to batches, but it seemed to make some developers nervous :)

**Comment by herbelin** [comment]
> IMHO it is up to authors and developers to signal that they want more discussion (using labels).

OK, I may try this and preventively set a `needs: discussion` on PRs about which I may have something to say.

**Comment by Zimmi48** [comment]
@ejgallego Can you make explicit what else you expect for this issue to be closed again?

**Comment by ejgallego** [comment]
We are having some problems with the effectiveness of the filter, IIUC. What should be done on that front?

**Comment by Zimmi48** [comment]
Can you elaborate? I'm not aware of these problems.

**Comment by ejgallego** [comment]
So for example, restarting builds don't remove them from the filter, it seems.

**Comment by ejgallego** [comment]
Also CI seems to hinder severely the merging ability.

**Comment by Zimmi48** [comment]
The CI is a big issue I agree but there are specific issues about this. Right now pending builds are filtered out so restarting a build should remove a PR from the filter. Failed builds are not filtered out because there are too many of them but if we manage to make the CI better, we could also filter out failed builds.

**Comment by ejgallego** [comment]
So I leave it to @maximedenes if he want to close this, or to propose an alternate method as to mark ready-to-merge PRs.

**Comment by ejgallego** [comment]
I am glad we didn't close this as the issue seems far to be solved. Opaqueness has indeed very much improved, however, we face new challenges on predictability. Two points:
- as of the time of writing this comment, out of 163 PR, 106 remain unassigned. That is not looking good.
- maintainer reactivity is still an open question. I am not sure all the maintainers will be reactive, and we lack a good procedure to workaround an inactive maintainer.

**Comment by maximedenes** [comment]
> a good procedure to workaround an inactive maintainer

I think we should first try to get active maintainers. We still have some freedom in changing assignments of maintainers.

**Comment by gares** [comment]
I think we need tools to measure. Our personal experience is too biased imo. Eg. I try to promptly set needs:review but I ve no way to be notified when things change, or when a timeout has expired. Maybe gh API + jq is enough

**Comment by Zimmi48** [comment]
Good point @gares! I'll try to set up some measurements and some bots to help in the coming months.

**Comment by ejgallego** [comment]
I am personally happy with the outcome of the change in the merging policy thus closing this issue as I don't feel the issue is gonna contribute nothing more.

PR merging has become much more predicatable as maintainers tend to be either responsive or MIA, and I guess individual subsystems should be addressed in individual issues such as #7345 #7346 .

On the other hand, the problem is still far from solved due to unassigned PRs and some other PRs [such as #6156 , #6958, #6959, etc........] being stuck for no discernible reason.

**Comment by SkySkimmer** [comment]
>PR merging has become much more predicatable as maintainers tend to be either responsive or MIA

MIA is certainly predictable but it's still pretty opaque.

https://github.com/coq/coq/pull/6959 has a pending request for changes.
